### PR TITLE
Pass headers

### PR DESCRIPTION
**Pull request by shrey-garg**

### CODE REVIEW

File: api/pom.xml [hunk 1]
```diff
@@ -10,24 +10,24 @@
     <parent>
         <groupId>com.flipkart.poseidon</groupId>
         <artifactId>poseidon</artifactId>
-        <version>3.1.1</version>
+        <version>4.0.0-SNAPSHOT</version>
     </parent>
 
     <artifactId>api</artifactId>
-    <version>3.1.1</version>
+    <version>${parent.version}</version>
```
Comment (pmohankumar) [inline api/pom.xml]: Just like groupId, we can ignore version too for child module. It should use the parent version

File: container/src/main/java/com/flipkart/poseidon/api/HeaderConfiguration.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright 2015 Flipkart Internet, pvt ltd.
```
Comment (pmohankumar) [inline container/src/main/java/com/flipkart/poseidon/api/HeaderConfiguration.java]: 2016

File: container/src/main/java/com/flipkart/poseidon/api/HeaderConfiguration.java [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+        return defaultValue;
+    }
+
+    public void setDefaultValue(String defaultValue) {
+        this.defaultValue = defaultValue;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
```
Comment (pmohankumar) [inline container/src/main/java/com/flipkart/poseidon/api/HeaderConfiguration.java]: Null check for o before doing instance of etc

File: container/src/main/java/com/flipkart/poseidon/api/Configuration.java [hunk 4]
```diff
@@ -42,4 +42,6 @@
     public JettyConfiguration getJettyConfiguration();
 
     public ExceptionMapper getExceptionMapper();
+
+    public Headers getConfiguredHeaders();
```
Comment (pmohankumar) [inline container/src/main/java/com/flipkart/poseidon/api/Configuration.java]: getConfiguredHeaders() doesn't sound right. Do you want to change it to getProxyHeadersConfiguration() or something like that?

File: container/src/main/java/com/flipkart/poseidon/api/Headers.java [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+
+package com.flipkart.poseidon.api;
+
+import java.util.List;
+
+/**
+ * Created by shrey.garg on 06/04/16.
+ */
+public interface Headers {
+    List<HeaderConfiguration> getPassThroughHeaders();
```
Comment (pmohankumar) [inline container/src/main/java/com/flipkart/poseidon/api/Headers.java]: Set is a natural fit for having unique headers?

File: service-clients-core/src/main/java/com/flipkart/poseidon/serviceclients/AbstractServiceClient.java [hunk 6]
```diff
@@ -135,17 +134,12 @@
             injectedHeadersMap.putAll(headersMap);
         }
 
-        // Add x-request-id
-        String requestId = (String) RequestContext.get(REQUEST_ID);
-        if (requestId != null && !requestId.isEmpty()) {
-            injectedHeadersMap.put(REQUEST_ID_HEADER, requestId);
+        // Inject Configured Headers
+        if (ServiceContext.get(HEADERS) instanceof Map) {
```
Comment (pmohankumar) [inline service-clients-core/src/main/java/com/flipkart/poseidon/serviceclients/AbstractServiceClient.java]: null check?

File: container/src/main/java/com/flipkart/poseidon/core/PoseidonServlet.java [hunk 7]
```diff
@@ -148,20 +151,26 @@ protected void doRequest(HttpMethod method, HttpServletRequest httpRequest, Http
 
     private void setRequestContext(HttpServletRequest httpServletRequest) {
         RequestContext.set(METHOD, httpServletRequest.getMethod());
-        RequestContext.set(REQUEST_ID, getRequestId(httpServletRequest));
-        RequestContext.set(IS_PERF_TEST, isPerfTest(httpServletRequest));
```
Comment (pmohankumar) [inline container/src/main/java/com/flipkart/poseidon/core/PoseidonServlet.java]: Do we need to remove request id and is perf test from existing RequestContext. May be we should populate both ServiceContext and RequestContext with the configured headers.

**Comment by pmohankumar** [comment]
Can you please create a wiki page for migration notes. Say users migrating from 3.x to 4.x have to provide:
1. Configuration.java getProxyHeadersConfiguration()
2. RequestContext.get(IS_PERF_TEST) and RequestContext.get(REQUEST_ID) should be replaced with ....

File: container/src/main/java/com/flipkart/poseidon/core/PoseidonServlet.java [hunk 8]
```diff
@@ -148,20 +151,26 @@ protected void doRequest(HttpMethod method, HttpServletRequest httpRequest, Http
...
         RequestContext.set(SOURCE_ADDRESS, httpServletRequest.getRemoteAddr());
     }
 
-    private String getRequestId(HttpServletRequest httpServletRequest) {
-        if (httpServletRequest.getHeader(REQUEST_ID_HEADER) != null) {
```
Comment (pmohankumar) [inline container/src/main/java/com/flipkart/poseidon/core/PoseidonServlet.java]: If REQUEST_ID_HEADER, REQUEST_ID, IS_PERF_TEST, PERF_TEST_HEADER constants are not used, can you remove them?

File: service-clients-core/src/main/java/com/flipkart/poseidon/serviceclients/AbstractServiceClient.java (hunk 6 above)
Comment (shrey-garg) [inline service-clients-core/src/main/java/com/flipkart/poseidon/serviceclients/AbstractServiceClient.java]: null instanceOf Map will be false

File: container/src/main/java/com/flipkart/poseidon/core/PoseidonServlet.java (hunk 7 above)
Comment (shrey-garg) [inline container/src/main/java/com/flipkart/poseidon/core/PoseidonServlet.java]: I have added a field in HeaderConfiguration, asking if the particular header is required in RequestContext

**Comment by shrey-garg** [comment]
https://github.com/flipkart-incubator/Poseidon/wiki/Migration-Guide-%5B3.x-to-4.x%5D

**Pull request merged**